Support INSERT in Google Sheets connector (#15026)
trinodb/trino: docs/src/main/sphinx/connector/googlesheets.rst

  <img src="../_static/img/google-sheets.png" class="connector-logo">

> The Google Sheets connector allows reading and writing `Google Sheets <https://www.google.com/sheets/about/>`_ spreadsheets as tables in Trino.

Reviewer: @mosabua Do you want to review this docs change?

Reviewer 2: @mosabua ping

Author: Thankfully he has already reviewed the docs

``metadata-sheet-id``               Sheet ID of the spreadsheet, that contains the table mapping
``sheets-data-max-cache-size``      Maximum number of spreadsheets to cache, defaults to ``1000``
``sheets-data-expire-after-write``  How long to cache spreadsheet data or metadata, defaults to ``5m``
``connection-timeout``              Timeout when connection to Google Sheets API, defaults to ``5m``
``read-timeout``                    Timeout when reading from Google Sheets API, defaults to ``5m``

> ``write-timeout``                   Timeout when writing to Google Sheets API, defaults to ``5m``

Reviewer: Please rename config properties:
* connection-timeout → gsheets.connection-timeout
* read-timeout → gsheets.read-timeout
* write-timeout → gsheets.write-timeout

--------------

The metadata sheet is used to map table names to sheet IDs.
The first row must be a header row containing the following columns in this order (the actual header value does not matter):

.. list-table:: Metadata sheet columns
   :widths: 25 45 30
   :header-rows: 1

   * - Column
     - Description
     - Default
   * - ``Table Name``
     - Name of the table as it will show up in Trino
     - no default - mandatory
   * - ``Sheet ID``
     - Sheet ID of the Google sheet. See `Sheet ID`_
     - no default - mandatory
   * - ``Owner``
     - Documentation purpose only
     - empty string
   * - ``Notes``
     - Documentation purpose only
     - empty string
   * - ``Column types``
     - Specify types of the columns. See `Column types`_

>      - no type mapping - all columns will be ``VARCHAR``

Reviewer: This change shouldn't belong to "Support INSERT in Google Sheets" commit. Please confirm other docs change as well.

Set the ``metadata-sheet-id`` configuration property to the ID of this sheet.

Sheet ID
^^^^^^^^
Sheet ID of the Google sheet.
You can extract this from the URL when editing a sheet in a browser.
Given the URL https://docs.google.com/spreadsheets/d/1eKgxd877iah7rFvtcphqSU1z5eA3eYKHKsYkDmErbwo/edit#gid=0, the sheet id is ``1eKgxd877iah7rFvtcphqSU1z5eA3eYKHKsYkDmErbwo``.

As a default, this connector will fetch 10,000 rows from the first tab in the sheet.
You need to prepend the tab name after the sheet ID separated with ``#`` in one of the following cases:

* Read more than 10,000 rows
* Read from a different tab withing the sheet

This could for example look like ``1Es4HhWALUQjoa-bQh4a8B5HROz7dpGMfq_HbfoaW5LM#Customer Orders``.

Column types
^^^^^^^^^^^^
Specify types of the columns.

> The format is ``mycol1=varchar,mycol2=bigint,mycol3=bigint``.

Reviewer 3: I think the issue is that the type assigned to a cell can change for every cell as you move down a column. 
But for actually querying stuff we ideally want a single type per column.

> The format is ``mycol1=varchar,mycol2=bigint,mycol3=bigint``.

Author: The updated PR only handles INSERTs, support for type mapping was dropped, hence the mapping is not needed any more

The columns names need to be called the same way as they show up in the Trino table, which is lowercase.
If no type is specified for a column or no column types are specified at all, ``varchar`` will be used.
The supported column types are documented in `Type mapping`_.

Querying sheets
---------------

The service account user must have access to the sheet containing the data in order for Trino to query it.

> Click the *Share* button to share the sheet with the email address of the service account.

Author: There actually is a change in it. Happy to to revert back to the original sentence line break if desired.

> Click the *Share* button to share the sheet with the email address of the service account.

Reviewer: Please revert.

Querying sheets
---------------
The service account user must have access to the sheet containing the data in order for Trino to query it.
Click the *Share* button to share the sheet with the email address of the service account.

Writing to sheets
-----------------
The same way sheets can be queried, they can also be written by appending data to existing sheets.
In this case the service account user must also have **write** access to the sheet.

After data is written to a table, the table contents are removed from the cache described in `Caching`_.
If the table is accessed imitatively after the write, querying the Google Sheets API may not reflect the change yet.
In that case the old version of the table will be read and cached for the configured amount of time.
So it might take some time for the written changes to propagate properly.

Please keep in mind that the Google Sheets API has `usage limits <https://developers.google.com/sheets/api/limits>`_, which limit the speed of inserting data.
You might run into timeouts (which you can increase using the configuration properties) or ``503: The service is currently unavailable`` errors.

> Caching

Reviewer: Revert unrelated change.

names of the Trino table.
For more details on sheet range syntax see the `google sheets docs <https://developers.google.com/sheets/api/guides/concepts>`_.

Writing to sheets
-----------------

> The same way sheets can be queried, they can also be written by appending data to existing sheets.

Reviewer 4: Empty line after title

In this case the service account user must also have **write** access to the sheet.

After data is written to a table, the table contents are removed from the cache described in `API usage limits`_.

> If the table is accessed imitatively after the write, querying the Google Sheets API may not reflect the change yet.

Reviewer 4: ```suggestion
If the table is accessed immediately after the write, querying the Google Sheets API may not reflect the change yet.
```

> In that case the old version of the table will be read and cached for the configured amount of time.

Reviewer 4: ```suggestion
In that case the old version of the table are read and cached for the configured amount of time, 
```

> So it might take some time for the written changes to propagate properly.

Reviewer 4: ```suggestion
and it might take some time for the written changes to propagate.
```

> Please keep in mind that the Google Sheets API has `usage limits <https://developers.google.com/sheets/api/limits>`_, which limit the speed of inserting data.

Reviewer 4: ```suggestion
Keep in mind that the Google Sheets API has `usage limits <https://developers.google.com/sheets/api/limits>`_, that limit the speed of inserting data.
```

> You might run into timeouts (which you can increase using the  `Configuration properties`_) or ``503: The service is currently unavailable`` errors.

Reviewer 4: ```suggestion
If you run into timeouts you can increase times to avoid ``503: The service is currently unavailable`` errors.
```

names of the Trino table.
For more details on sheet range syntax see the `google sheets docs <https://developers.google.com/sheets/api/guides/concepts>`_.

Writing to sheets
-----------------

The same way sheets can be queried, they can also be written by appending data to existing sheets.

> In this case the service account user must also have **write** access to the sheet.

Reviewer 2: I thought that Google Docs is using `edit` term instead of `write`.

SQL support
-----------

> In addition to the :ref:`globally available <sql-globally-available>` and :ref:`read operation <sql-read-operations>` statements,

Reviewer 4: Empty line after title

this connector supports the following features:

* :doc:`/sql/insert`

> * :doc:`/functions/table`

Reviewer 4: Lets leave this out here since this is not done in any other connector doc either.

Author: Removed the `* :doc:/functions/table` section
